### DISCUSSION TITLE
Using Fore with WebSockets in BaseX

### DISCUSSION THREAD

**Discussion by timathom** [Jinntec/Fore#208] [Q&A] [answered]
BaseX is my back-end environment of choice for XRX applications, and it [supports the WebSockets protocol](https://docs.basex.org/wiki/WebSockets). I have yet to make use of this functionality in an application, but I'm wondering: could it work with Fore (for example, in a chat app like the demo included in the BaseX distribution)?

**Comment by JoernT** [comment, marked as answer]
That's an interesting question! Actually WebSocket have never been integrated with the XForms Spec. and currently there's no support for them in Fore. 

However i don't see a reason why we couldn't have it and certainly that's an interesting one. It's more a question of how to sensibly wrap that into a component. fx-submission doesn't feel like the right term at first sight. However the term 'submission' is not a very lucky choice anway as it implies and emphasizes more the 'sending' part though it very well also handles the response or receiving part. 

Beyond terminology questions submission and websockets are similar:
* they connect to a URL
* they are used to send/receive data
* they have some eventing for success/error
* they need a way to be triggered - (fx-send for submission)

So, first question for me would be if it should be a new component like fx-websocket, fx-channel (suggestions?) or should we try to extend fx-submission? Thinking about it maybe its more sensible to go with the first as we'll need some extra events and submit-done, submit-error also don't feel appropriate to cover a WebSocket opening/closing.

Preliminary Pseudo Code:
```
...
<fx-model>
<fx-connection id="chat" url="wss://URL:PORT" ref="xpath-to-connected-data" protocol="chat, message">
   <fx-message event="connection-open">We have been connected</fx-message>
   <fx-message event="connection-error">We had a bad start</fx-message>
   <fx-message event="connection-closed">We have been disconnected</fx-message>
   <fx-message event="data-received">We received data<fx-message>
<fx-connection>
...
</fx-model>
```
`protocol` is to conform to WebSockets API and allow to specifiy the subprotocol(s).

Thinking aloud about a possible implementation:
* on page load (during model-construct?)  fx-connection would try to connect (do we need to trigger that explicitly?) and establish a server connection. If that's going well we would get a 'connected' event. Otherwise we'd get 'connection-error'. We can message users appropriately.
* From now on when we receive data from the server we replace the bound node(s) with the data received and may bind them in our UI as usual. This will trigger an update cycle as necessary.
* To send data we could actually re-use/extend the fx-send action maybe like `<fx-send connection="chat">`

Open Questions:
* WebSockets have no defined data format so can transport txt, xml, json whatever. We could determine the desired data-format from the bound data or adding an attribute to fx-connection. 
* for completeness we should call close() on the connection at some time. Do we need to support an explicit fx-close-connection? Or should that be handled transparently by the processor? The latter assumes that the connection is always living as long as the page session.

As you might notice - my interest has awakened ;) Would really like to see that - however, no promises on timelines yet.

**Reply by timathom** [reply]
Glad to see you are interested! :) I agree this would be a very cool feature to have in Fore. 

`fx-connection` sounds like a good name--agree that "submission" doesn't make sense in this context. I think it makes sense to connect on page load, but it might also be useful to be able to trigger a connection on demand.

I notice in the [BaseX demo app](https://github.com/BaseXdb/basex/blob/bed209fcb9127663fd20552c8767880256c29d00/basex-api/src/main/webapp/static/chat.js#L5), there's a JS function that's triggered `onopen` to send a heartbeat to keep the connection alive, so would be good if that could be handled (or maybe have an attribute to set the value, with a default).

**Reply by timathom** [reply]
And in XForms, the [`submission` element](https://www.w3.org/community/xformsusers/wiki/XForms_2.0#The_submission_Element) has a `mediatype` attribute, which might be repurposed here.

**Reply by JoernT** [reply]
update: i did a still rough implementation but it works insofar that i can exchange messages between browsers connected to the same WS. It will need more design and polishing with regard to the integration before rolling out.
